mono/metadata/android.c

```diff
@@ -0,0 +1,358 @@
+_monodroid__system_property_get (const gchar *name, gchar *sp_value, gsize sp_value_len)
+{
+	if (!name || !sp_value)
+		return -1;
+
+	gchar *buf = NULL;
+	if (sp_value_len < PROP_VALUE_MAX + 1) {
+		g_warning ("Buffer to store system property may be too small, will copy only %u bytes", sp_value_len);
+		buf = xmalloc (PROP_VALUE_MAX + 1);
+	}
+
+	gint len = __system_property_get (name, buf ? buf : sp_value);
+	if (buf) {
+		strncpy (sp_value, buf, sp_value_len);
+		sp_value [sp_value_len] = '\0';
+		free (buf);
+	}
+
+	return len;
+}
+#endif
+
+static gint
+monodroid_get_system_property (const gchar *name, gchar **value)
+{
+	gchar *pvalue;
+	gchar  sp_value [PROP_VALUE_MAX+1] = { 0, };
+	gint   len;
+	struct BundledProperty *p;
+
+	g_assert (value);
+	*value = NULL;
+
+	pvalue = sp_value;
+	len = _monodroid__system_property_get (name, sp_value, sizeof (sp_value));
+
+	if (len <= 0 && (p = lookup_system_property (name)) != NULL) {
+		pvalue = p->value;
+		len = p->value_len;
+	}
+
+	if (len >= 0)
+		*value = g_strndup (pvalue, len);
+	return len;
+}
+
+gpointer
+ves_icall_System_TimeZoneInfo_AndroidTimeZones_GetDefaultTimeZoneId (MonoError *error)
+{
+	error_init (error);
+
+	JNIEnv *env         = ensure_jnienv ();
+	jobject d           = (*env)->CallStaticObjectMethod (env, TimeZone_class, TimeZone_getDefault);
+	jstring id          = (*env)->CallObjectMethod (env, d, TimeZone_getID);
+	const gchar *mutf8  = (*env)->GetStringUTFChars (env, id, NULL);
+
+	gchar *def_id = g_strdup (mutf8);
+
+	(*env)->ReleaseStringUTFChars (env, id, mutf8);
+	(*env)->DeleteLocalRef (env, id);
+	(*env)->DeleteLocalRef (env, d);
+
+	return def_id;
+}
+
+gpointer
+ves_icall_System_TimeZoneInfo_AndroidTimeZones_GetPersistTimeZoneId (MonoError *error)
+{
+	gchar *id;
+	gint ret;
+
+	error_init (error);
+
+	ret = monodroid_get_system_property ("persist.sys.timezone", &id);
+	if (ret <= 0)
+		return NULL;
+
+	return id;
+}
+
+MONO_API void
+monodroid_free (gpointer ptr)
+{
+	g_free (ptr);
+}
+
+#if __arm__
+static gint
+find_in_maps (const gchar *str)
+{
+	FILE  *maps;
+	gchar *line;
+	gchar  buf [512];
+
+	g_assert (str);
+
+	maps = fopen ("/proc/self/maps", "r");
+	if (!maps)
+		return -1;
+
+	while ((line = fgets (buf, sizeof(buf), maps))) {
+		if (strstr (line, str)) {
+			fclose (maps);
+			return 1;
+		}
+	}
+
+	fclose (maps);
+	return 0;
+}
+#endif
+
+static guchar
+is_64_bit ()
+{
+	return sizeof (gchar*) == 8;
+}
+
+#define CPU_KIND_UNKNOWN ((gushort)0)
+#define CPU_KIND_ARM     ((gushort)1)
+#define CPU_KIND_ARM64   ((gushort)2)
+#define CPU_KIND_MIPS    ((gushort)3)
+#define CPU_KIND_X86     ((gushort)4)
+#define CPU_KIND_X86_64  ((gushort)5)
+
+static gint
+get_built_for_cpu_android (gushort *built_for_cpu)
+{
+	gint retval = 1;
+
+#if __arm__
+	*built_for_cpu = CPU_KIND_ARM;
+#elif __aarch64__
+	*built_for_cpu = CPU_KIND_ARM64;
+#elif __x86_64__
+	*built_for_cpu = CPU_KIND_X86_64;
+#elif __i386__
+	*built_for_cpu = CPU_KIND_X86;
+#elif __mips__
```

Comment:
should this use `TARGET_ARM`/`TARGET_ARM64`/... etc instead?

Reply:
It most likely should, let me fix it.